### PR TITLE
Setup websockets communication between web and workers

### PR DESCRIPTION
**Pull request by andresgutgon**
# What?
We want to keep the client informed when evaluations are running. How many logs are created, and how many logs fail
This PR set up a new Express server that set up a Socket.IO websocket server with 2 endpoints /websocket and worker-websocket. The one for the web has secure cookies token auth and the one for workers share a secret token between Sockets server and workers.

I think this setup makes sense

https://github.com/user-attachments/assets/c38bd897-7871-42ee-ac0c-1945c2562a10

## TODO
- [x] Setup Sockets server
- [x] Setup setting token and refresh token cookies in `web` when user login with magic link.
- [x] Setup Sockets client in worker so it can emit events to the websockets server
- [x] Setup DOMAIN_NAME env variable.
- [x] Setup `<WebsocketsProvider />` in React client so it can start listening to events
- [x] Make Websocket connection work `WEB`
- [x] Make Websocket connection work `WORKERS`
- [x] Run evaluation and see if it triggers a websocket to the frontend
- [x] Build UI to show evaluation progress.
- [x] WHY first time a evaluation is run `total` is 0? REVIEW this
- [x] Setup refresh websocket token server action
- [x] Review ENV variables to avoid breaking deploy with this PR
- [x] Implement `logout` lol we didn't have it
- [x] Delete Websocket cookies when `logout`

## Next PR
- [x] Review all necessary env variables
- [ ] Think about `ws` or `wss` (Secure). Maybe is not necessary because is behind a proxy cc @geclos 
- [ ] Talk about private/public keys by workspace cc @geclos 
- [ ] @socket.io/redis-adapter for pub/sub with Redis. Maybe? I don't think we need now if it works without it cc @geclos

### CODE REVIEW

File: packages/jobs/src/job-definitions/batchEvaluations/runEvaluationJob.ts [hunk 1]
```diff
@@ -45,6 +45,18 @@ export const runEvaluationJob = async (job: Job<RunEvaluationJobData>) => {
     }
 
     await progressTracker.incrementErrors()
+  } finally {
     await progressTracker.decrementTotal()
```
Comment (andresgutgon) [inline packages/jobs/src/job-definitions/batchEvaluations/runEvaluationJob.ts:49]: decrementTotal was only the catch but I think is something we want to do always no?

File: apps/web/package.json [hunk 2]
```diff
@@ -27,10 +27,12 @@
     "@sentry/nextjs": "^8",
     "@t3-oss/env-nextjs": "^0.10.1",
     "ai": "^3.2.42",
+    "socket.io-react-hook": "^2.4.5",
```
Comment (andresgutgon) [inline apps/web/package.json:30]: Not very complicated to integrate but this package is small and nice

File: apps/web/package.json [hunk 3]
```diff
@@ -42,6 +43,8 @@
     "nprogress": "^0.2.0",
     "react": "19.0.0-rc-f994737d14-20240522",
     "react-dom": "19.0.0-rc-f994737d14-20240522",
+    "socket.io-client": "^4.7.5",
```
Comment (andresgutgon) [inline apps/web/package.json:46]: 🔥 Included in the hook

File: apps/websockets/package.json [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    "cookie-parser": "^1.4.6",
+    "express": "^4.21.0",
+    "socket.io": "^4.7.5"
+  },
+  "devDependencies": {
+    "@latitude-data/eslint-config": "workspace:^",
+    "@latitude-data/typescript-config": "workspace:^",
+    "@types/cookie-parser": "^1.4.7",
+    "@types/express": "^4.17.21",
+    "@types/jsonwebtoken": "^9.0.6",
```
Comment (andresgutgon) [inline apps/websockets/package.json:28]: 🔥

File: apps/workers/src/server.ts [hunk 5]
```diff
@@ -6,7 +6,7 @@ const workers = setupWorkers()
 
 console.log('Workers started')
 
-const port = 3000
+const port = 3001
```
Comment (andresgutgon) [inline apps/workers/src/server.ts:9]: This was conflicting with apps/web @geclos

File: packages/env/src/index.ts [hunk 6]
```diff
@@ -66,9 +71,18 @@ export const env = createEnv({
...
+    APP_DOMAIN: z.string(),
+    APP_SUBDOMAIN: z.string().optional(),
   },
   runtimeEnv: {
     ...process.env,
+    WEBSOCKETS_SERVER:
+      process.env.WEBSOCKETS_SERVER ?? 'https://ws.latitude.so',
+    APP_DOMAIN: process.env.APP_DOMAIN ?? 'latitude.so',
+    APP_SUBDOMAIN: process.env.APP_SUBDOMAIN ?? 'app',
     FILE_PUBLIC_PATH: process.env.FILE_PUBLIC_PATH ?? FILE_PUBLIC_PATH,
```
Comment (andresgutgon) [inline packages/env/src/index.ts:86]: FAKE SECRETS in this PR. Not necessary until we deploy it

File: packages/core/src/websockets/utils.ts [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  const secret = new TextEncoder().encode(env.WEBSOCKET_SECRET_TOKEN_KEY)
+  const token = await new SignJWT(payload)
+    .setProtectedHeader({ alg: 'HS256' })
+    .setExpirationTime(config.maxAge.stringValue)
+    .sign(secret)
+  return {
+    token,
+    cookiesOptions: {
+      secure: isProd,
+      domain: isProd ? `.${env.APP_DOMAIN}` : 'localhost',
```
Comment (andresgutgon) [inline packages/core/src/websockets/utils.ts:25]: CORS for all the subdomains of the app domain. This way we can share cookies between web client and websockets server as auth method

File: apps/web/src/env.ts [hunk 8]
```diff
@@ -23,5 +24,6 @@ export default createEnv({
     GATEWAY_HOSTNAME: process.env.GATEWAY_HOSTNAME,
     GATEWAY_PORT: process.env.GATEWAY_PORT,
     GATEWAY_SSL: process.env.GATEWAY_SSL,
+    WEBSOCKETS_SERVER: process.env.WEBSOCKETS_SERVER,
```
Comment (geclos) [inline apps/web/src/env.ts:27]: no hay default aqui

File: apps/websockets/src/server.ts [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+const isProd = process.env.NODE_ENV === 'production'
+const server = http.createServer(app)
+const subdomain = env.APP_SUBDOMAIN ? `${env.APP_SUBDOMAIN}.` : ''
+const io = new Server(server, {
+  path: '/websocket',
+  cors: {
+    origin: isProd
+      ? `https://${subdomain}${env.APP_DOMAIN}`
+      : 'http://localhost:3000',
```
Comment (geclos) [inline apps/websockets/src/server.ts:47]: env var instead of isprod check would be cooler

File: apps/websockets/src/server.ts [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+
+const app = express()
+app.use(cookieParser())
+app.get('/health', (_, res) => {
+  res.json({ status: 'Websockets server running' })
+})
+
+const isProd = process.env.NODE_ENV === 'production'
+const server = http.createServer(app)
+const subdomain = env.APP_SUBDOMAIN ? `${env.APP_SUBDOMAIN}.` : ''
```
Comment (geclos) [inline apps/websockets/src/server.ts:41]: we have LATITUDE_URL and LATITUDE_DOMAIN env vars can we work with those?

File: packages/jobs/src/job-definitions/batchEvaluations/runEvaluationJob.ts (hunk 1 above)
Comment (geclos) [reply packages/jobs/src/job-definitions/batchEvaluations/runEvaluationJob.ts:54]: hmm no we only wanna do it when there's an error, if we succeed we don't wanna decrement the total

File: apps/web/src/env.ts (hunk 8 above)
Comment (andresgutgon) [reply apps/web/src/env.ts:27]: It's on `packages/env/src/index.ts`

**Review by geclos — APPROVED** [review]

**Pull request merged**